### PR TITLE
[BC-breaking] Fixed interpolation output size to match opencv if scale factor is specified

### LINKED ISSUE #62396: Make interpolation output size compatible with opencv, scikit-image and scipy for floating scale factor
## 🚀 Feature

Currently, interpolation output size if using float scale factor is computed as 
https://github.com/pytorch/pytorch/blob/c06b6e445f022b062924ae9b49e2739c2992d067/aten/src/ATen/native/UpSample.cpp#L24

This can lead to -1 pixel output size comparing to opencv, scikit-image or scipy. For example, let's rescale an image of size (32, 32) with different scale factors: `[0.1234, 0.9999, 1.8]`. This gives the following output sizes per library:
```
- check scale_factor=0.1234
Error: different shapes: opencv (4, 4) vs sk-img (4, 4) vs scipy (4, 4) vs pth (3, 3)

- check scale_factor=0.9999
Error: different shapes: opencv (32, 32) vs sk-img (32, 32) vs scipy (32, 32) vs pth (31, 31)

- check scale_factor=1.8
Error: different shapes: opencv (58, 58) vs sk-img (58, 58) vs scipy (58, 58) vs pth (57, 57)

...
```

Here is a gist with details: https://gist.github.com/vfdev-5/a5bd5b1477b1c82a87a0f9e25c727664

## Pitch

It would be helpful to have compatible output size using pytorch interpolate op.

## Alternatives

Use output size argument directly instead of scales.

## Additional context

Here is equivalent code for 
- opencv: 
  - https://github.com/opencv/opencv/blob/68d15fc62edad980f1ffa15ee478438335f39cc3/modules/imgproc/src/resize.cpp#L4049-L4058
  - https://github.com/opencv/opencv/blob/68d15fc62edad980f1ffa15ee478438335f39cc3/modules/core/include/opencv2/core/saturate.hpp#L140
- scikit-image: https://github.com/scikit-image/scikit-image/blob/e128e0743bb2cded104654352a13ed5b4e53efb8/skimage/transform/_warps.py#L286
- scipy: https://github.com/scipy/scipy/blob/44e31b1eb202ef91caad059952ba268d2610a60b/scipy/ndimage/interpolation.py#L768-L769

cc @fmassa

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #62396

Dscription:
- Fixed interpolation output size to match opencv, scikit-image, scipy if scale factor is specified

cc @heitorschueroff , @fmassa

### CODE REVIEW

**Review by fmassa — APPROVED** [review]
LGTM, thanks!

This is a BC-breaking change though, so I would love to get @jbschlosser approval on how to move forward with the breaking change.

**Comment by vfdev-5** [comment]
@jbschlosser could you please take a look at this PR which should fix output size issue by torch.nn.functional.interpolation vs opencv, scikit-image.

**Review by jbschlosser — COMMENTED** [review]
@fmassa do you have a sense for how disruptive this change would be if we made it outright? The breakage seems less common and less silent than fixing the off-by-one interpolate content error, and it's still possible to get the old output by slicing (e.g. `out[:,:,:-1,:-1]`) if necessary.

That said, the conservative route I'd default to involves the painful process of adding another flag to switch between old and new behavior (idk a good name, `round_with_scale_factor`?), the [JIT FC workaround](https://github.com/pytorch/pytorch/issues/67565#issuecomment-961999987) for adding the new kwarg, etc. If the breakage isn't too bad, I'd like to avoid this messy situation, but it's hard for me to say.

**Comment by vfdev-5** [comment]
> The breakage seems less common and less silent than fixing the off-by-one interpolate content error, and it's still possible to get the old output by slicing (e.g. out[:,:,:-1,:-1]) if necessary.

@jbschlosser I think this can be a bit more trickier, for example https://github.com/pytorch/pytorch/issues/70559 where slicing does not give the previous output.

**Comment by jbschlosser** [comment]
> @jbschlosser I think this can be a bit more trickier, for example #70559 where slicing does not give the previous output.

Good point and thanks for bringing it up - I think we'll have to go the long and painful route. I did a quick github code search and it does look like the behavior will be disruptive for certain use cases, especially for things like U-Nets which have a lot of `scale_factor` usage.

I'll put forth the name of `round_with_scale_factor` for the new kwarg, but other suggestions are welcome. Suggested deprecation process, following that done for interpolate previously:
1. Introduce new kwarg (e.g. `round_with_scale_factor=None`) for a release that acts disabled by default. Users can explicitly opt-in to the fix. If no explicit value is specified for this flag AND the resultant shape will be calculated differently, warn that the default behavior will change in a future release. Users can explicitly set a value for the kwarg to acknowledge and turn off the warning.
2. Change the default behavior for the kwarg to be enabled and remove the warning. At this point, users have had a full release to explicitly choose between rounding and no rounding. The bug fix will be applied by default, and anyone who knows they do not want rounding for some reason can explicitly disable it.

@albanD / @mruberry wdyt about the above?

**Comment by albanD** [comment]
Sounds good to me!

**Comment by vfdev-5** [comment]
Sorry for delay on this PR. I still plan to follow https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629 and update the PR. I remove "Stale" label.

**Comment by NicolasHug** [comment]
Just chiming in as @vfdev-5 asked me to take a look at this. The strategy described in https://github.com/pytorch/pytorch/pull/64983#issuecomment-1026177629 sounds like the least disruptive way forward. Perhaps with a small amendment: according to https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy "stage 1" should probably last 2 versions instead of 1.
